### ISSUE TITLE
How to use System.Net.Http 4.3.3 with .NET 4.6.0

### ISSUE THREAD

**Issue by kasper3** [dotnet/corefx#25050]
Took a unit test from https://github.com/dotnet/corefx/pull/19082 that fixed the issue of reusing content.

UnitTest1.cs:

```
using System;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using Xunit;

public class UnitTest1
{
    [Fact]
    public async Task PostAsync_ReuseRequestContent_Success()
    {
        var assemblyName = typeof(HttpClient).Assembly.GetName();
        Console.WriteLine($"{assemblyName.Name} version is: {assemblyName.Version}");

        var remoteServer = new Uri("http://httpbin.org/post");
        const string ContentString = "This is the content string.";
        using (var client = new HttpClient())
        {
            var content = new StringContent(ContentString);
            for (int i = 0; i < 2; i++)
            {
                using (HttpResponseMessage response = await client.PostAsync(remoteServer, content))
                {
                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
                    Assert.Contains(ContentString, await response.Content.ReadAsStringAsync());
                }
            }
        }
    }
}
```

mytest.csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net46</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.3.0" />
    <PackageReference Include="xunit" Version="2.3.1" />
    <PackageReference Include="System.Net.Http" Version="4.3.3" />
  </ItemGroup>

  <ItemGroup>
    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
  </ItemGroup>

</Project>
```

Then ran:
```
dotnet restore
dotnet xunit
```

Result (throws second iteration):

```
System.Net.Http version is: 4.1.1.2
    UnitTest1.PostAsync_ReuseRequestContent_Success [FAIL]
      System.ObjectDisposedException : Cannot access a disposed object.
      Object name: 'System.Net.Http.StringContent'.
      Stack Trace:
           at System.Net.Http.HttpContent.CheckDisposed()
           at System.Net.Http.HttpContent.CopyToAsync(Stream stream, TransportContext context)
           at System.Net.Http.HttpClientHandler.GetRequestStreamCallback(IAsyncResult ar)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
```

that means it's using old assembly (first output line: System.Net.Http version is: 4.1.1.2)

Then replaced:

```
    <TargetFramework>net46</TargetFramework>
```

with:

```
    <TargetFramework>netcoreapp2.0</TargetFramework>
```

restored and ran xunit command again

Result:

```
System.Net.Http version is: 4.2.0.0
  Finished:   mytest
=== TEST EXECUTION SUMMARY ===
   mytest  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.673s
```

Two questions:

1. How can I ***force*** framework 4.6.0 to use System.Net.Http from the Reference package?
2. (if binding redirect is the only way) how can I add binding redirect in new csproj format?

**Comment by davidsh** [comment]
>Took a unit test from #19082 that fixed the issue of reusing content.

That change about "no longer disposing HttpRequestMessage.Content" was only done for .NET Core.  That change is not in .NET Framework.  So, running this test on .NET Framework will fail as you observed.

**Comment by kasper3** [comment]
Does that mean:

A: using System.Net.Http 4.3.0 from nuget on .NET Framework 4.6.0 will not hit that code path in HttpClient

or

B: there is no way to run System.Net.Http 4.3.0 on NET Framework 4.6.0, it will always pick old one form GAC

Looking at the code changes i would assume it would be `B`

**Comment by davidsh** [comment]
>Looking at the code changes i would assume it would be B

It's actually more complicated than that.

Starting with .NET Framework 4.7.1, we will always prefer to use the inbox System.Net.Http.dll (GAC'd version) and not any OOB (out-of-band) System.Net.Http.dll anymore from NuGet package.  Due to various issues described in #17522, we no longer build OOB assemblies for System.Net.Http.dll.  As of .NET 4.7.1, the API surface for System.Net.Http.dll matches NetStandard20/NetCore20.

In your example, there will be ways you could potentially force the Http.dll binary to be used from the NuGet package.  But that will only work for a while until you run on .NET 4.7.1.  After that, there is no separate dll that would be used from NETCore sources/NuGet packages against .NET Framework.  With respect to the behavior change ("not explicitly disposing HttpRequestMessage.Content after sending message"), that change has not bee ported to .NET Framework yet (need to review app-compat mitigations).

In summary, concepts like this:

>using System.Net.Http 4.3.0 from nuget on .NET Framework

won't be true moving forward.

cc: @karelz

**Comment by kasper3** [comment]
Thanks for the explanation davidsh. Much appreciated.

**Comment by sjnorval** [comment]
I see that this issue has been closed although I don't seem to get how to fix the problem... Please help:
![image](https://user-images.githubusercontent.com/4039677/32605493-4d4b5f22-c55a-11e7-963b-e30bc4ca3b3a.png)

I am using .netFramework 4.7 and I have the following nuget package installed:
![image](https://user-images.githubusercontent.com/4039677/32605570-921a35ce-c55a-11e7-8679-245bf609b2f2.png)

**Comment by buddyspike** [comment]
@davidsh is above explanation valid for .Net framework 4.6 as well?

**Comment by davidsh** [comment]
>@davidsh is above explanation valid for .Net framework 4.6 as well?

@joperezr Can you comment on this?

**Comment by joperezr** [comment]
Yes, System.Net.Http.dll also shipped inbox on .NET Framework 4.6 so because of our conflict resolution logic, that one will be preferred.
